src/Proxy/ProxyGenerator.php

```diff
@@ -0,0 +1,91 @@
+<?php
+
+namespace Zenstruck\Foundry\Proxy;
+
+use Doctrine\ORM\EntityManagerInterface;
+use ProxyManager\Factory\AccessInterceptorValueHolderFactory;
+use Zenstruck\Foundry\Configuration;
+
+/**
+ * Generates model proxies to autorefresh property values
+ * from the database.
+ *
+ * @author Wouter de Jong <wouter@wouterj.nl>
+ */
+class ProxyGenerator
+{
+    /** @var Configuration */
+    private $configuration;
+    /** @var AccessInterceptorValueHolderFactory */
+    private $factory;
+
+    public function __construct(Configuration $configuration, ?AccessInterceptorValueHolderFactory $factory = null)
+    {
+        $this->configuration = $configuration;
+        $this->factory = $factory ?? new AccessInterceptorValueHolderFactory();
+    }
+
+    public function generate(object $object, array $methods = []): object
+    {
+        $objectManager = $this->configuration->objectManagerFor(\get_class($object));
+        $interceptor = function(object $proxy, object $model) use ($objectManager): void {
+            $autoRefreshEnabled = \property_exists($model, '_foundry_autoRefresh') ? $model->_foundry_autoRefresh : true;
+            if (!$autoRefreshEnabled) {
+                return;
+            }
+
+            $modelClassMetadata = $objectManager->getClassMetadata(\get_class($model));
+
+            // only check for changes if the object is managed in the current om
+            if ($objectManager instanceof EntityManagerInterface && $objectManager->contains($model)) {
+                // cannot use UOW::recomputeSingleEntityChangeSet() here as it wrongly computes embedded objects as changed
+                $objectManager->getUnitOfWork()->computeChangeSet($modelClassMetadata, $model);
+
+                if (!empty($objectManager->getUnitOfWork()->getEntityChangeSet($model))) {
+                    throw new \RuntimeException(\sprintf('Cannot auto refresh "%s" as there are unsaved changes. Be sure to call ->save() or disable auto refreshing (see https://github.com/zenstruck/foundry#auto-refresh for details).', \get_class($model)));
+                }
+
+                $objectManager->refresh($model);
+
+                return;
+            }
+
+            // refetch the model as it's no longer managed
+            $modelId = $modelClassMetadata->getIdentifierValues($model);
+            if (!$modelId) {
+                // TODO: throw exception?
```

Comment:
The current `Proxy` class sets the value holder to `null`. Is that what we want? Or should we throw an exception? Or leave the value holder as-is?

Reply:
If I understand correctly, currently, if the object cannot be "refreshed", we [throw an exception](https://github.com/zenstruck/foundry/blob/d2cd4c7b4c6821b96d8fda0c1d2febad345ffa35/src/Proxy.php#L167).

Reply:
Oh, I misread the code. Exception it is :)

> By "value holder", do you mean this? 

Yes, in ProxyManager terms, "value holder" is the object that is being proxied.